content/iaed/1000-algoritmos.sheet.md

```diff
@@ -20,45 +20,381 @@ O código disponível nesta página foi retirado de slides dos docentes, bibliog
+}
+
+Item aux[DIM];
+
+void merge(Item a[], int left, int m, int right) {
+  int i, j, k;
+  for (i = m + 1; i > left; i--)
+    aux[i - 1] = a[i - 1];
+  for (j = m; j < right; j++)
+    aux[right + m - j] = a[j + 1];
+  for (k = left; k <= right; k++) {
+    // vamos escolhendo os elementos das pontas para ordenar o vetor
+    if (less(aux[j], aux[i]) || i == m + 1)
+      a[k] = aux[j--];
+    else
+      a[k] = aux[i++];
+  }
+}
+```
+
+:::
 
 ### Heap Sort
 
-`embed:assets/1000-heap.c`
+[Explicação nos resumos](../algoritmos-eficientes-ordenacao#heap-sort-enquadramento).
+
+:::details[Heap Sort]
+
+```c
+void heapsort(Item a[], int l, int r) {
+  buildheap(a, l, r);
+  while (r - l > 0) {
+    exch(a[l], a[r]);
+    fixDown(a, l, --r, l);
+  }
+}
+
+void buildheap(Item a[], int l, int r) {
+  int k, heapsize = r - l + 1;
+  for (k = heapsize / 2 - 1; k >= l; k--)
+    fixDown(a, l, r, l + k);
+}
+
+void fixDown(Item a[], int l, int r, int k) {
+  int ileft, iright, largest = k;
+  ileft = l + left(k - l);
+  iright = l + right(k - l);
+
+  if (ileft <= r && less(a[largest], a[ileft]))
+    largest = ileft;
+  if (iright <= r && less(a[largest], a[iright]))
+    largest = iright;
+
+  if (largest != k) {
+    exch(a[k], a[largest]);
+    fixDown(a, l, r, largest);
+  }
+}
+
+int parent(int k) {
+  return ((k + 1) / 2) - 1;
+}
+
+int left(int k) {
+  return 2 * k + 1;
+}
+
+int right(int k) {
+  return 2 * (k + 1);
+}
+```
+
+:::
 
 ### Counting Sort
 
-`embed:assets/1000-counting.c`
+[Explicação nos resumos](../algoritmos-elementares-ordenacao-2#counting-sort).
+
+**De realçar que a implementação do _counting sort_ varia dependendo da situação**: o que é intrínseco à lógica acaba por manter-se, mas temos obviamente que utilizar este algoritmo para ordenar caracteres e inteiros será diferente. A implementação abaixo corresponde a uma implementação-exemplo para ordenação de um vetor de inteiros.
+
+:::details[Counting Sort]
+
+```c
+#define MAX 100 // 100 é placeholder
+/* vamos evitar o uso de variáveis na declaração de vetores,
+ * porque variable length arrays are bad
+ * devemos OU usar constantes OU usar malloc e amigos
+ * wikipedia article relevante:
+ * https://en.wikipedia.org/wiki/Variable-length_array#C99
+ */
+
+void countingSort(int numbers[], int size) {
+  int count[SIZE];
+  for (int i = 0; i < MAX; i++)
+    count[i] = 0;
+
+  for (int i = 0; i < size; i++)
+    count[numbers[i]]++;
+
+  int index = 0;
+  for (int i = 0; i < MAX; i++) {
+    while (count[i] > 0) {
+      numbers[index] = i;
+      index++;
+      count[i]--;
+    }
+  }
+}
+
+```
+
+:::
 
 ### Radix Sort LSD
 
-`embed:assets/1000-radixLSD.c`
+[Explicação nos resumos](../algoritmos-elementares-ordenacao-2#radix-lsd).
+
+:::details[Radix Sort LSD]
+
+```c
+#define maxN 100000
+#define bitsword 32
+#define bitsbyte 8
+#define bytesword 4
+#define R (1 << bitsbyte)
+#define digit(n, w) ((n & ( 1 << w )) >> w)
+
+// defines um pouquinho cursed, mas é o que há
```

Comment:
Vê o que tenho mesmo nos resumos, provavelmente é melhor

Reply:
Fixed in the next commit